### PR TITLE
Tandem Difficulty Curriculum: progressive exposure by gap/stagger magnitude

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model sees all training samples with equal probability throughout training. But tandem configurations with extreme gap and stagger values are much harder — p_tan is 2.37x harder than p_in. A model trained on all difficulty levels simultaneously must balance competing gradient signals from easy (single-foil) and hard (extreme-gap tandem) samples, which may cause the easy samples to dominate early training and push the backbone into a basin that generalizes poorly to hard tandem cases.

Curriculum learning (Bengio et al., 2009) proposes starting with easier samples and progressively introducing harder ones. This is a DATA SCHEDULING approach — no architecture changes, pure training schedule. Orthogonal to all other modifications.

**Key bet:** Progressive tandem exposure lets the model first build a solid foundation on easy/medium tandem configs, then adapt to extreme gap/stagger cases. This directly targets p_tan by ensuring the model sees a coherent difficulty gradient rather than random difficulty at every batch.

**This is NOT a repeat of `tandem_ramp`** (already merged): tandem_ramp modulates the LOSS WEIGHT for the tandem task gradient. This experiment modulates WHICH tandem samples appear in each batch — they're complementary, non-overlapping mechanisms.

**Literature:**
- Bengio et al. "Curriculum Learning" (ICML 2009)
- Hacohen & Weinshall "On The Power of Curriculum Learning" (ICML 2019) — curriculum helps most for hard OOD test cases
- Platanios et al. "Competence-based Curriculum Learning" (NAACL 2019)

### Step 1: Add Curriculum Scheduler

Add a `TandemDifficultyScheduler` class in train.py:

```python
class TandemDifficultyScheduler:
    """Progressive tandem sample curriculum by gap+stagger magnitude."""
    def __init__(self, tandem_indices, gap_values, stagger_values,
                 warmup_epochs=40, full_epochs=80):
        self.warmup = warmup_epochs
        self.full = full_epochs
        # Rank tandem samples by difficulty: normalized (|gap| + |stagger|)
        difficulty = np.abs(gap_values) + np.abs(stagger_values)
        difficulty = (difficulty - difficulty.min()) / (difficulty.max() - difficulty.min() + 1e-6)
        self.sorted_indices = tandem_indices[np.argsort(difficulty)]
        self.n_tandem = len(tandem_indices)

    def get_available_indices(self, epoch):
        """Return tandem indices available at this epoch."""
        if epoch < self.warmup:
            # Easiest 25% only
            n_avail = max(1, self.n_tandem // 4)
        elif epoch < self.full:
            # Linear ramp from Q1 to all
            frac = (epoch - self.warmup) / (self.full - self.warmup)
            n_avail = int(self.n_tandem * (0.25 + 0.75 * frac))
        else:
            n_avail = self.n_tandem
        return self.sorted_indices[:n_avail]
```

### Step 2: Integrate with DataLoader

In the training loop's DataLoader/sampler:
1. At each epoch, call `scheduler.get_available_indices(current_epoch)` to get the set of tandem sample indices to include
2. Replace the uniform tandem sampling with this filtered set
3. Non-tandem samples (single-foil, extreme-Re) are ALWAYS sampled uniformly from the full set — the curriculum only applies to tandem configurations
4. The gap and stagger values for each tandem sample should be available from the dataset metadata (TandemFoilSet stores these). If not directly accessible, compute from the sample's geometry.

### Step 3: New Flags

- `--tandem_curriculum` (bool, default False) — enable curriculum scheduling
- `--curriculum_warmup_epochs 40` (int) — epochs using only easiest 25% of tandem samples
- `--curriculum_full_epochs 80` (int) — epoch at which all tandem samples become available (linear ramp from warmup to full)

### Step 4: Key Interactions with Existing Stack

- **PCGrad**: still applies — tandem task gradient remains separate from single-foil/Re tasks
- **tandem_ramp**: still applies at the sample level — the curriculum operates at batch-composition level, orthogonal
- **Augmentation**: still applies to all samples including curriculum-selected tandem samples
- **gap_stagger_spatial_bias**: still active — the spatial bias features are computed for whatever tandem samples are in the batch

### Step 5: Training Runs

Run 2 seeds:

```
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/tandem-difficulty-curriculum-s42" \
  --wandb_group "tandem-difficulty-curriculum" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --tandem_curriculum --curriculum_warmup_epochs 40 --curriculum_full_epochs 80
```

Then seed 73 with same flags but `--seed 73`.

### Important Notes

- **Visualize first:** Before running, print the gap/stagger distribution to confirm there's a meaningful difficulty gradient. If all tandem samples cluster tightly, the curriculum has no structure to exploit.
- **Epoch budget matters:** With warmup=40 and full=80, the model sees only Q1 tandem samples for the first 40 epochs. Ensure the Q1 set has enough samples for meaningful batches — if TandemFoilSet has very few tandem samples total, reduce warmup_epochs proportionally.
- **EMA interaction:** EMA continues updating from epoch 0. In the early curriculum phase, EMA captures the "easy-tandem" model — this is fine, the EMA will update through the full curriculum.

## Baseline

Current best metrics (PR #2251, 2-seed avg):
- p_in: **11.891** (target: < 11.89)
- p_oodc: **7.561** (target: < 7.56)
- p_tan: **28.118** (target: < 28.12)
- p_re: **6.364** (target: < 6.36)
- W&B baseline runs: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Metrics (Surface MAE pressure, lower is better)

| Metric | Baseline (2-seed avg) | Curriculum s42 | Curriculum s73 | Curriculum 2-seed avg | Δ vs baseline |
|--------|----------------------|----------------|----------------|----------------------|---------------|
| p_in   | 11.891               | 12.362         | 12.355         | 12.358               | +0.467 (+3.9%)  |
| p_oodc | 7.561                | 8.493          | 7.999          | 8.246                | +0.685 (+9.1%)  |
| p_tan  | 28.118               | 30.883         | 30.246         | 30.564               | +2.447 (+8.7%)  |
| p_re   | 6.364                | 6.593          | 6.804          | 6.699                | +0.335 (+5.3%)  |

### Curriculum progression

| Epoch range | Tandem samples available | Fraction |
|-------------|-------------------------|----------|
| 0-39        | 189 / 756               | 25%      |
| 40-79       | 189→756 (linear ramp)   | 25%→100% |
| 80+         | 756 / 756               | 100%     |

### Run details

- **W&B run IDs**: `g8jwdot4` (s42), `cztl9tw2` (s73)
- **W&B group**: `tandem-difficulty-curriculum`
- **Epochs**: 155 (both seeds) — hit 180-min timeout
- **Best epoch**: 154 (both seeds) — late convergence, possibly still improving at cutoff
- **Peak VRAM**: ~46 GB (same as baseline — no architecture changes)
- **Training overhead**: ~5 min one-time curriculum scan at startup; per-epoch sampler weight update is negligible

### Command used

```bash
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/tandem-difficulty-curriculum-s42" \
  --wandb_group "tandem-difficulty-curriculum" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --tandem_curriculum --curriculum_warmup_epochs 40 --curriculum_full_epochs 80
```

### What happened

**Negative result.** The tandem difficulty curriculum degraded all 4 metrics. The worst impact was on p_tan (+8.7%), the very metric the curriculum was designed to improve. p_oodc also degraded significantly (+9.1%).

### Analysis

1. **Reduced tandem exposure during critical early training**: By restricting to only 25% of tandem samples for the first 40 epochs, the model saw far fewer tandem configurations during the critical early learning phase. The baseline benefits from uniform sampling across all tandem configs from epoch 0 — this diversity helps the model generalize from the start. The curriculum effectively withheld 75% of tandem training data during the period when the model is most receptive to learning new patterns.

2. **Interaction with `tandem_ramp` is counterproductive**: The existing `tandem_ramp` already implements a gradual introduction mechanism — it ramps the tandem **loss weight** from 0→1 over epochs 10-50. Combining curriculum (restricting which samples) with tandem_ramp (restricting how much we learn from them) creates a double-gating effect. During warmup, the model sees only easy tandem samples AND learns from them with reduced weight. This severely under-trains the tandem pathway.

3. **Late convergence suggests insufficient exposure**: Both seeds had best_epoch=154, much later than the baseline (typically ~120-130). This suggests the model was still trying to catch up on tandem configurations introduced late in training. The curriculum delayed when the model could see hard tandem samples, but those hard samples are precisely the ones that need the most training iterations.

4. **Gap/stagger difficulty may not align with learning difficulty**: The curriculum ranks difficulty by |gap| + |stagger| magnitude, but the model's actual learning difficulty for tandem samples depends on complex interactions (Reynolds number, angle of attack, wake interference) that don't correlate linearly with gap+stagger magnitude. Easy-to-predict tandem configs might have large gap/stagger values and vice versa.

### Suggested follow-ups

- **Anti-curriculum**: Instead of withholding hard samples, try UP-WEIGHTING hard tandem samples from the start. Loss-based difficulty (actual prediction error) would be a better difficulty signal than geometric features.
- **Shorter warmup**: If curriculum is worth retrying, warmup=10 and full=30 would minimize the exposure gap while still providing some ordering benefit.
- **Error-based curriculum**: Rank tandem difficulty by actual validation error rather than geometric features. This would require periodic re-ranking but would better capture actual learning difficulty.
- **Remove tandem_ramp when using curriculum**: The double-gating effect is clearly harmful. If trying curriculum again, disable tandem_ramp to avoid compounding the exposure reduction.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2277 because ADVISOR: Closing — all 4 metrics regressed (p_in +3.9%, p_oodc +9.1%, p_tan +8.7%, p_re +5.3%).

**Root cause:** Excellent analysis. Two compounding issues: (1) The curriculum withholds 75% of tandem samples during epochs 0-40 — precisely when the model is most receptive to learning new patterns. The baseline benefits from full tandem diversity from epoch 0. (2) tandem_ramp already gates tandem learning (loss weight 0→1 over epochs 10-50). Adding curriculum creates double-gating that severely under-trains the tandem pathway.

**Key takeaway:** The existing uniform sampling + tandem_ramp is already a well-tuned exposure schedule. Restricting sample diversity during early training hurts more than the ordered difficulty gradient helps. The model needs to see hard tandem samples early to allocate capacity for them. Added to DO NOT REVISIT.

Good suggestion about anti-curriculum/error-based weighting. Moving you to a different direction..
